File: Daily/Rust/Leetcode1600-皇位继承顺序.rs
use std::collections::{HashMap, HashSet};
struct ThroneInheritance {
    king:String,
    relation:HashMap<String,Vec<String>>,
    dead:HashSet<String>
}
impl ThroneInheritance {

    fn new(kingName: String) -> Self {
        Self{
            king:kingName,
            relation:HashMap::new(),
            dead:HashSet::new()
        }
    }
    
    fn birth(&mut self, parent_name: String, child_name: String) {
        self.relation.entry(parent_name).or_insert(Vec::new()).push(child_name)
    }
    
    fn death(&mut self, name: String) {
        self.dead.insert(name);
    }
    
    fn get_inheritance_order(&self) -> Vec<String> {
        let mut ans = Vec::new();
        self.pre_order(&self.king, &mut ans);
        ans
    }
    
    fn pre_order(&self, name: &String, ans: &mut Vec<String>) {
        if !self.dead.contains(name) {
            ans.push(name.to_string());
        }
        
        self.relation.get(name).unwrap_or(&vec![]).iter().for_each(|child| self.pre_order(child, ans));
    }
}

File: Leetcode/sbgo/Leetcode1600-皇位继承顺序.go
package sbgo

type ThroneInheritance struct {
	king     string
	relation map[string][]string
	dead     map[string]bool
}

func Constructor(kingName string) ThroneInheritance {
	return ThroneInheritance{kingName, map[string][]string{}, map[string]bool{}}
}

func (t *ThroneInheritance) Birth(parentName string, childName string) {
	t.relation[parentName] = append(t.relation[parentName], childName)
}

func (t *ThroneInheritance) Death(name string) {
	t.dead[name] = true
}

func (t *ThroneInheritance) GetInheritanceOrder() []string {
	ans := []string{}
	var dfs func(string)
	dfs = func(name string) {
		if !t.dead[name] {
			ans = append(ans, name)
		}
		for _, childName := range t.relation[name] {
			dfs(childName)
		}
	}
	dfs(t.king)
	return ans
}
